src/gameengine/graphics/Renderer.java

```diff
@@ -0,0 +1,355 @@
+package gameengine.graphics;
+
+import gameengine.geometry.Vector2D;
+
+import java.awt.*;
+
+/**
+ * The Renderer is responsible for drawing on the current frame.
+ */
+public abstract class Renderer {
+    /**
+     * Sets the foreground color.
+     *
+     * @param color The foreground color
+     */
+    public final void setForegroundColor(MutableColor color) {
+        setForegroundColor(color.getRed(), color.getGreen(), color.getBlue());
+    }
+
+    /**
+     * Sets the foreground color.
+     *
+     * @param red   The red component
+     * @param green The green component
+     * @param blue  The blue component
+     */
+    public abstract void setForegroundColor(float red, float green, float blue);
+
+    /**
+     * Sets the background color.
+     *
+     * @param color The background color
+     */
+    public final void setBackgroundColor(MutableColor color) {
+        setBackgroundColor(color.getRed(), color.getGreen(), color.getBlue());
+    }
+
+    /**
+     * Sets the background color.
+     *
+     * @param red   The red component
+     * @param green The green component
+     * @param blue  The blue component
+     */
+    public abstract void setBackgroundColor(float red, float green, float blue);
+
+    /**
+     * Controls which elements are drawn in front of others.  Elements with a greater index are
+     * drawn behind elements with a smaller index.
+     *
+     * @param index the Z index
+     */
+    public abstract void setZIndex(float index);
+
+    /**
+     * Sets the scaling factor.
+     *
+     * @param scale The scaling factor
+     */
+    public abstract void setScale(double scale);
```

Comment:
Hmm, I was going to say sure but this adds some complication.  My current OpenGL implementation sets the screen co-ordinates at the beginning of each frame by calculating them based on the scale but it looks like you are thinking of changing the scale several times in 1 frame.

I was thinking about a simple use-case where you can zoom with the mouse wheel.  In OpenGL, you just draw an image and specify the coordinates which controls the size.

One way to add this capability in OpenGL would be to modify all coordinates for every single draw call based on the scale that we would store and this would slow things down.  But... OpenGL must have some sort of transformation matrix that I could modify so it seems doable.

This brings us back to the ordering problem.  If we want to zoom with the mouse wheel, the scale needs to be set before anything is drawn on the screen so I guess the viewport class can handle that.

Ok, I'm fairly convinced.  I'll make the change :)